rewriter/flatten.cc

```diff
@@ -82,19 +82,18 @@ class FlattenWalk {
         //  def self.bar; end
         //
         // which means when we get to `bar` we need to know that the outer context `foo` is static. We pass that down
-        // the current stack by means of this `staticLevel` variable.
-        int staticLevel;
-        MethodData(optional<int> targetLocation, int staticLevel)
-            : targetLocation(targetLocation), staticLevel(staticLevel){};
+        // the current stack by means of this `isStatic` variable.
+        bool isStatic;
```

Comment:
Having a simple boolean here raises my eyebrow. 
is it possible to write a testcase that has a singleton-of-singleton and thus it's a method o f
```
class <<self
  class <<self
     def foo
     end
  end
end
```

Reply:
I'll add this test-case, but the changes I'm making here won't affect how this test-case behaves because it doesn't contain any nested method definitions. Once you're inside a method body, you're either inside a static method body (in which case `def foo` defines an instance method and `def self.foo` defines a static method) or you're inside an instance method (in which case `def foo` defines another instance method but `def self.foo` defines an instance method that is only defined on a specific instance), and no more granular tracking is necessary as long as all you're doing is un-nesting method definitions.

Reply:
sorry, I didn't mean exact this case, i'm rather asking if there's a case that should flatten into this.

Reply:
It's worth noting that we have a test already that includes a lot of `class << self`—specifically, [desugar/sclass.rb](https://github.com/sorbet/sorbet/blob/master/test/testdata/desugar/sclass.rb)—and that the changes in this PR didn't change any of the expectations for it.

Reply:
Oh, I did misunderstand: AFAIK, there's no way with just nesting of method defs to create anything other than static methods, instance methods, or instance-methods-that-are-limited-to-a-single-instance. If you nest static methods, like this
```ruby
class A
  def self.foo
    def self.bar
      def self.baz
        def self.quux
        end
      end
    end
  end
end
```
they'll all just end up being static methods on `A` (i.e. `A.foo`, `A.bar`, `A.baz`, and `A.quux`.)

Reply:
It turns out flatten doesn't currently correctly handle
```ruby
def foo
  class << self
    def bar; end
  end
end
```
so I'm going to fix that in another PR for simplicity.